### DISCUSSION TITLE
ESP32-S3 : Hitting upper limit of open endpoints when I should not; issue in portable/synopsys/dcd_dwc2.c?

### DISCUSSION THREAD

**Discussion by giri-a** [hathach/tinyusb#2627] [Q&A] [answered]
I am implementing a headset (speaker+mic) on ESP32-S3. I notice that when I pause the playback or record 3 times, playback or record does not happen any more.
I traced the issue to a couple of TU_ASSERT in dcd_edpt_counter_release(). When I pause playback, the endpoint is released and opened again when I start playing.

```
// Release an endpoint.
static bool dcd_edpt_counter_release(uint8_t rhport, uint8_t dir)
{
  if (dir) {
    TU_ASSERT(dwc2_ep_counters[rhport].active_in == 0);    // Check if number of opened EPs is not zero
    dwc2_ep_counters[rhport].active_in--;                  // Decrease endpoint counter
  } else {
    TU_ASSERT(dwc2_ep_counters[rhport].active_out == 0);
    dwc2_ep_counters[rhport].active_out--;
  }
  return true;
}
```

Obviously the ASSERT fails when active_in and active_out are >0 and they are not decremented.  Do this 3 times and total number of open endpoints are 6, which the limit. Subsequently, dcd_edpt_counter_available() returns false.  

When I changed '==' to '>', the problem went away. It makes sense to me. Also the comment implies that it should be '!=' or '>'.

Anybody else faced this issue? Am I missing something or thinking it wrong?

**Comment by HiFiPhile** [comment]
Wrong repo, `dcd_edpt_counter_release()` is not a part of tusb stack, you should ask expressif.

**Reply by HiFiPhile** [reply, marked as answer]
Or you can use tusb without expressif wrapper, check how examples are built.

**Reply by giri-a** [reply]
My bad! I should have asked there.
